WalletWasabi/Crypto/StrobeProtocol/StrobeHasher.cs

```diff
@@ -0,0 +1,57 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+using NBitcoin;
+using WalletWasabi.WabiSabi;
+
+namespace WalletWasabi.Crypto.StrobeProtocol
+{
+	public sealed class StrobeHasher
+	{
+		private Strobe128 strobe;
+
+		private StrobeHasher(string domain)
+		{
+			strobe = new Strobe128(ProtocolConstants.WabiSabiProtocolIdentifier);
+			Append(ProtocolConstants.DomainStrobeSeparator, domain);
+		}
+
+		public void Append(string fieldName, object fieldValue)
+			=> Append(fieldName, fieldValue switch
+			{
+				IBitcoinSerializable serializable => serializable.ToHex(),
+				Money money => money.Satoshi.ToString(CultureInfo.InvariantCulture),
+				string str => str,
+				uint numUint => numUint.ToString(CultureInfo.InvariantCulture),
+				ulong numUlong => numUlong.ToString(CultureInfo.InvariantCulture),
+				decimal numDecimal => numDecimal.ToString(CultureInfo.InvariantCulture),
+				CredentialIssuerParameters issuerParameters => issuerParameters.ToString(),
+				_ => throw new ArgumentException($"{fieldValue.GetType().Name} doesn't have a string representation for strobe hasher.")
+			});
+
+		public void Append(string fieldName, string fieldValue)
+		{
+			strobe.AddAssociatedMetaData(Encoding.UTF8.GetBytes(fieldName), false);
+
+			var serializedValue = Encoding.UTF8.GetBytes(fieldValue);
+			strobe.AddAssociatedMetaData(BitConverter.GetBytes(serializedValue.Length), true);
+			strobe.AddAssociatedData(serializedValue, false);
+		}
```

Comment:
Here's what I was thinking. This is nicer than a switch because it's statically typed and should never generate a runtime exception, it's more efficient than canonicalizing to strings, and leaves no room for misinterpretation.

The fieldnames implicitly commit to the type and STROBE delimitation means we don't need to encode the length.

```suggestion
		public void Append(string fieldName, IBitcoinSerializable serializable) => Append(fieldName, serializable.ToBytes());
		public void Append(string fieldName, Money money) => Append(fieldName, money.Satoshis())
		public void Append(string fieldName, uint numUint) => Append(fieldName, /* just encode with known endianness */)
		public void Append(string fieldName, ulong numUlong) => Append(fieldName, /* just encode with known endianness */)
		public void Append(string fieldName, CredentialIssuerParameters issuerParameters => Append(fieldName, issuerParameters.ToBytes()); // just concatenate the group elements 

		public void Append(string fieldName, decimal numDecimal) => Append(fieldName, numDecimal.ToString(CultureInfo.InvariantCulture)) // is it really needed?
		// TimeSpan?

		private void Append(string fieldName, byte[] serializedValue)
		{
			strobe.AddAssociatedMetaData(Encoding.UTF8.GetBytes(fieldName), false);

			strobe.AddAssociatedMetaData(BitConverter.GetBytes(serializedValue.Length), true);
			strobe.AddAssociatedData(serializedValue, false);
		}
```

Reply:
This was my first idea but what to do if the type if not one of those? in Ruby you have a catch-all mechanism but in c# there is none. 
The alternative is remove the `Combine` method.

Reply:
nothing, just let the compiler error, if the comitted values' types changes that's a protocol breaking thing so we should avoid leaving it open for extension, and it's better to not wait until runtime to throw the error

Reply:
Done. You were right, it is much better now. Thank you.